### DISCUSSION TITLE
Include dependency to DKAN in contrib module info.yml

### DISCUSSION THREAD

**Discussion by stefan-korn** [GetDKAN/dkan#4149] [Q&A]
How can I include a dependency to DKAN (or even a submodule like metastore) in a info.yml for a contributed module?

I can write dependency in info.yml like on of this:
```
dependencies:
  - dkan:metastore
  - metastore:metastore
  - drupal:metastore
```

This will work if the module is only a local module. But if it is contrib module the Drupal composer packaging does somehow translate the dependencies from info.yml to composer dependencies, thus composer require drupal/custom_module will require dependency like drupal/metastore. But it cannot find drupal/metastore in the drupal composer packages (because DKAN is on Github), resulting in composer message like this

`drupal/metastore * -> could not be found in any version, there may be a typo in the package name.`

This seems to be the case no matter how the dependency is written in the info.yml. Drupal seems to do some magic here: https://drupal.stackexchange.com/questions/302656/how-is-composer-including-a-dependent-module

Does anybody know a solution to this problem?

PS: I can include dependency to DKAN in composer.json of the contrib module. But this will only composer install DKAN but does not have an effect on the order of module installation in Drupal.

**Comment by paul-m** [comment]
Drupal's Composer facade does a lot of magic in order to discover submodules with in the composer package, so it requires the correct composer package. (As an example, if you say `composer require drupal/file_example` Drupal's composer repo will install `drupal/examples`.) However, this doesn't happen for external dependencies, such as DKAN.

If you're making a Drupal-hosted contrib module that needs DKAN, you should add a composer.json file to it that requires `getdkan/dkan`. This gets you the entire DKAN package, including metastore. Just treat it like any other external dependency: https://www.drupal.org/docs/develop/using-composer/managing-dependencies-for-a-contributed-project

If the contrib module needs a specific DKAN module to be enabled at runtime, then as you point out above, use the info.yml to declare dependencies, so that Drupal will then try to enable that module on install.

**Comment by stefan-korn** [comment]
@paul-m : thanks for your answer.

I get it to work like you described making composer require getdkan/dkan with composer.json in the contrib module.

But doing this:

> If the contrib module needs a specific DKAN module to be enabled at runtime, then as you point out above, use the info.yml to declare dependencies, so that Drupal will then try to enable that module on install.

does always end up in composer searching for drupal/metastore (for example) and not finding this. The Drupal magic seems to add every dependency in the modules info.yml as a composer dependency as well, but always searching only in the "drupal" composer namespace.

**Reply by paul-m** [reply]
Last I heard, if you supply a composer.json file in your contrib module, the drupal composer repo won't try to add it from info.yml. That may have changed since I looked into it though.

**Reply by stefan-korn** [reply]
There are issues at the [Composer facade](https://www.drupal.org/project/project_composer) that seem to address this too, like:
https://www.drupal.org/project/project_composer/issues/2864379
https://www.drupal.org/project/project_composer/issues/3304661

But I get no clue from this right now. I am not sure if I can formulate the info.yml dependency to not have Composer facade look in drupal/..., tried like "module", "drupal;module", "module:module", "package:module", but all seem to end up looking for drupal/module in composer.

My issue was with installing DKAN from a profile and getting the wrong installation order of contributed module dependening on DKAN and DKAN itself. Now I solved this by installing the contributed module depending on DKAN in hook_install in the profile. This solves it for me now, because this way the contributed module surely comes after DKAN.

But in general there seem to be an issue in this process. Find it kind of opaque that Drupal is transforming the info.yml dependencies in composer dependencies. Would be cleaner to just rely on composer.json or at least to not look in info.yml if there is a composer.json in the module like you proposed or supposed.

**Comment by swirtSJW** [comment]
@stefan-korn Now that select2 package dependency has been removed from the dkan core composer.json with release 2.21, you can now use DKAN and its submodules as a dependency both in your .info.yml and in composer.json

*.info.yml
```
dependencies:
  - dkan:dkan
  - dkan:common
  - dkan:datastore

```

composer.json 
```
```json
    "require": {
      "getdkan/dkan": "^2.21",
```"require": {
      "[getdkan/dkan](https://packagist.org/packages/getdkan/dkan)": "[^2.21](https://packagist.org/packages/getdkan/dkan)",
```

**Comment by stefan-korn** [comment]
@swirtSJW : thanks for your message. But I suppose this does not solve this specific problem about including dependency to DKAN in contrib module info.yml.

**Reply by swirtSJW** [reply]
@stefan-korn Tying this in with conversations you and I have been having here https://www.drupal.org/project/dkan_dataset_archiver/issues/3547877 
and
https://www.drupal.org/project/dkan_chart/issues/3547643
for the benefit of @dafeder, @janette and @paul-m 

I thought that release 2.21 had fixed the blocker of being able to set dkan core and its sub modules as dependencies but you are absolutely right, it only fixed half the problem.  Your sharp eye spotted the problem that lead me to think it was working. 
I thought it fixed the problem because when I did `composer require drupal/dkan_dataset_archiver` it seemed to work.  I did not catch that it was only pulling 1.0.0-alpha1 (which was before I included dkan dependencies in the info.yml) and not grabbing the most recent 1.0.0-alpha3.  
The failure is in the packaging done by Drupal.org, in the additional data that the packager provides.  It apparently does not know how to handle dependencies for items outside of drupal.org so it dies and does not create package info correctly and keeps it stalled at 1.0.0-alpha1


So far I think there are three possible solutions:

1. Move dkan to drupal.org.  This would solve all the problems (but is a bit of  a lift that in my opinion we should be doing sooner rather than later)
2. Find the problem witthin https://www.drupal.org/project/project_composer and fix it (also worth doing)
3. Short term solution for my module which breaks if the dependencies are not declared in the info.yml even if the modules are enabled, is to remove the dependencies from the info.yml AND include a .patch file with the module that can be added by hand to composer patches (requires sites to use  [cweagens composer patches](https://github.com/cweagans/composer-patches) that would put the dependencies back into the info.yml.

**Comment by stefan-korn** [comment]
@swirtSJW : Thanks for your summary and your proposals to solve this.

I would very much vote for No. 1, maybe it is possible to have only a mirror on drupal.org that does just provide releases and the code and issues stay on Github?

Regarding No 2: Would of course be great. But imho it is not so easy to grasp what is going on there even with looking in the code of project_composer, because this is also somehow tied to the drupal.org D7 instance. And even if a solution could be found easily, I doubt that it would find its way very soon in the code, as I suppose they will be very wary about changes to that project as this affects the whole drupal composer infrastructure. And I would not be wondering if they are already working on porting this to D8.x (D11) and would be just more reluctant regarding changes in the old project.

No. 3 sounds interesting. Will also have a try on this for dkan_chart probably. Although patching info.yml is sometimes difficult because Drupal is adding packaging information to the info.yml and that often interferes with a patch on dependencies as far as I know.

As another possible workaround I was thinking about enabling DKAN modules maybe through hook_install, but did not yet try this.

**Reply by swirtSJW** [reply]
@stefan-korn   

> As another possible workaround I was thinking about enabling DKAN modules maybe through hook_install, but did not yet try this.

That will likely not fix the problem. Yes the modules will enable but the contrib module would know know of the dependency which affects the dependency tree in drupal.  I know in the dkan_dataset_archiver some of my code fails if I do not have the dependency specified because not all the classes are loaded in the proper order so some are not there when I call on them.

@dafeder and I just spent some more time looking at this and the composer_packager.
The issue seems to be that when I declare a dependency on dkan:dkan it has no trouble because it finds [project/dkan](https://www.drupal.org/project/dkan).  The problem is that repository has no code.  Just a readme.  So it can't find dkan:common because common is not there to find.  It would also fail to find  datastore and metastore as well but it gives up before it tries them.

The solution will be to get the full codebase of dkan on drupal.org.  In order to do that we need to re-namespace the submodules because there is already a collision with https://www.drupal.org/project/datastore. After some testing I discovered that the dependencies don't know the difference between `- dkan:common` and `- drupal:common` The presence of either one satisfies the dependency and due to the previous paragraph it knows nothing of `dkan:common`.

I will put up a PR on dkan_dataset_archiver to do the patch info approach.  I am pretty sure I can get it to work... but I have been wrong before. :(

**Reply by stefan-korn** [reply]
Yes, true. Dependency in info.yml also has relevance for installation order etc. So my idea with hook_install is probably no good.

**Comment by swirtSJW** [comment]
I did get Solution #3 working as a temporary measure by including a patch with the module
You can see the composer changes here
[https://git.drupalcode.org/project/dkan_dataset_archiver/-/blob/1.0.x/co...](https://git.drupalcode.org/project/dkan_dataset_archiver/-/blob/1.0.x/composer.json?ref_type=heads)

and the patch and README here [https://git.drupalcode.org/project/dkan_dataset_archiver/-/tree/1.0.x/pa...](https://git.drupalcode.org/project/dkan_dataset_archiver/-/tree/1.0.x/patches?ref_type=heads)

If the site uses web as the docroot and already use composer-patches, then it works immediately.   If not it needs the patch to be added to the root composer by hand.
